### PR TITLE
Add support for well known types for buf beta convert

### PR DESCRIPTION
**Pull request by joshcarp**
Adds ability to call `buf beta convert` without specifying an input when a well known type is specified;

```
$ buf beta convert --type=google.protobuf.Duration --from <(echo \"3600s\")#format=json
> some proto bytes

$ buf beta convert --type=google.protobuf.Duration --from duration.bin
> "3600s"
```

### Input is defaulted to "."
This means that by default calling `buf beta convert --type=google.protobuf.Duration` will look for `google.protobuf.Duration` in the local directory first before resolving to a baked in well known type

Without explicit setting of input to be "." no errors will be reported on. This is to stop from `buf beta convert` failing to work in any directory that might have a broken proto file in. This means that:
- `buf beta convert .  --type=google.protobuf.Duration` will fail if `broken.proto` exists, for example but
- `buf beta convert --type=google.protobuf.Duration` will pass because the error from parsing `"."` will be discarded because the user did not specifically request for it to be parsed.

### CODE REVIEW

File: private/buf/bufcli/bufcli.go
Comment (jhump) [inline private/buf/bufcli/bufcli.go:812]: `protoparse` is the old compiler. Ideally, we'd figure out a way to make this work through the `bufimagebuild` package, since that consolidates all knowledge of the compiler (and also implements existing support for auto-discovery of the well known imports).

File: private/buf/buffetch/internal/reader.go
Comment (ElliotMJackson) [inline private/buf/buffetch/internal/reader.go:376]: extract method for early return to improve readability.

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (ElliotMJackson) [inline private/buf/cmd/buf/command/beta/convert/convert.go:238]:
```suggestion
	if !datawkt.Exists(filePath) {
		// not a well-known type
		return ""
	}
	// found it!
	return filePath
```

left align happy path

**Review by ElliotMJackson — CHANGES_REQUESTED** [review]

**Review by ElliotMJackson — APPROVED** [review]

File: private/buf/buffetch/internal/reader.go
Comment (bufdev) [inline private/buf/buffetch/internal/reader.go:366]: This means that you will always prioritize the built-in well-known types over whatever is on disk. As a counterexample of how this could break down, assume I have an old, incompatible version of `google/protobuf/timestamp.proto` on disk at `./google/protobuf/timestamp.proto`, and I specify `buf beta convert google/protobuf/timestamp.proto`. With this code, I will always use the built-in version and not the one on disk, whereas the rest of our code prioritizes the one on disk over the builtins.

File: private/buf/buffetch/internal/reader.go
Comment (bufdev) [inline private/buf/buffetch/internal/reader.go:367]: Why would I return the entire bucket if I have a single file? Shouldn't `buf beta convert google/protobuf/timestamp.proto --type google.type.Duration` fail?

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (bufdev) [inline private/buf/cmd/buf/command/beta/convert/convert.go:227]: I'm skeptical of this approach We should never be using the global registry for anything we do here. The global registry changes depending on what happens to be linked into the binary.

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (bufdev) [inline private/buf/cmd/buf/command/beta/convert/convert.go:159]: I believe this breaks the precedence rules for WKTs in input vs the builtins.

Also see our style guide, specifically the section on descriptive names, i.e. `wellKnownPath` would be correct here if this code persisted.

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (bufdev) [inline private/buf/cmd/buf/command/beta/convert/convert.go:224]: I can't comment a few lines up, but I'm a little surprised the `inverseEncoding` function is persisting to stable, I would have thought we removed this for stable, worth a discussion.

**Review by bufdev — CHANGES_REQUESTED** [review]

File: private/buf/buffetch/internal/reader.go
Comment (jhump) [reply private/buf/buffetch/internal/reader.go:367]: The compiler's source accessor has the logic for falling back when needed. So maybe the change needed is in the file provider for OS, below, to simply _allow_ "file not found" issues for paths that exist in `datawkt`. WDYT?

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (jhump) [reply private/buf/cmd/buf/command/beta/convert/convert.go:227]: I had suggested this as an alternative to a hand-written hard-coded map of type names to file names. The limitation you mention is easily worked around using underscore imports.

But another option would be to have a code generator (one that is run alongside `storage-to-go`) generate that hard-coded index by actually compiling the files in `datawkt` first and examining all of the types therein. That way it is not hand-coded and will never be inaccurate/stale.

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (joshcarp) [reply private/buf/cmd/buf/command/beta/convert/convert.go:224]: https://github.com/bufbuild/buf/issues/1432#issuecomment-1255555035

input encoding and output encoding can be explicitly set. inverse encoding only gets called if output doesn't have a `-#format` option (or the format can't be inferred by the filename)

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (joshcarp) [reply private/buf/cmd/buf/command/beta/convert/convert.go:159]: If the input is unspecified it will try to resolve the well known type, if not it will use the input argument. 

This means that calls like previously mentioned `buf beta convert google/protobuf/timestamp.proto --type google.type.Duration` will fail because `google/protobuf/timestamp.proto` doesn't contain `google.type.Duration`

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (jhump) [inline private/buf/cmd/buf/command/beta/convert/convert.go:181]: This seems odd. So if they provide inputs to identify the source of the well-known type, but they give the wrong inputs, we just ignore them and use the well-known type data? Seems like we should only fall back to the well-known type data when (1) no input files are named or (2) they name a file that is _in_ the well-known type data and _not_ found locally. Wouldn't both of those conditions result in `inputErr != nil` above?

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (jhump) [inline private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:178]:
```suggestion
		// stderr since we do output to stdout
```

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (jhump) [inline private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:186]: Is this right? Mysterious magic numbers and empty string args aren't good for readability. Let's at least add a comment if this is intentional.

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (jhump) [inline private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:191]: This seems like an ideal use for a template. I think that would be much more readable.

File: private/gen/data/datawkt/datawkt.gen.go
Comment (jhump) [inline private/gen/data/datawkt/datawkt.gen.go:68]: QQ: would this have been easier to just emit these to a separate generated file from the blob above? I naively would have thought we'd leave the existing generator alone for the data blob and just add another generated file...

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (joshcarp) [reply private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:191]: Personally I don't like text templates too much. As soon as there's any sort of control flow that needs to happen it becomes too complicated for me. 

I've changed all the `buffer.WriteString(` for an anonymous function `p(` which is similar to how it's done with other go code generators i've seen and it makes it a lot more readable than `buffer.WriteString`

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (joshcarp) [inline private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:150]: I am aware that this todo is here but I do not know what the difference between `bufmodule.NewModuleForBucket` and `bufmodulebuild.NewModuleBucketBuilder` is

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (jhump) [reply private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:191]: > As soon as there's any sort of control flow that needs to happen it becomes too complicated for me.

I agree 100%... which is why this is the ideal use for one since it's just emitting a hard-coded map 😝 

(Not a blocking comment... just my 2 cents on readability for this function.)

File: private/buf/cmd/buf/testdata/convert/bin_json/convert_test.go
Comment (bufdev) [inline private/buf/cmd/buf/testdata/convert/bin_json/convert_test.go:16]: You can do i.e. `buf convert testdata/convert/bin_json ...` correct? If so, if you move this test up a few levels, convert to this. If not, talk to me about why so (offline, moving off this PR after this round of comments)

File: private/bufpkg/bufwkt/cmd/wkt-go-data/main.go
Comment (bufdev) [reply private/bufpkg/bufwkt/cmd/wkt-go-data/main.go:193]: For context, this is mostly copy/pasted from `storage-go-data`, didn't want to spend time reinventing the wheel.

**Review by bufdev — APPROVED** [review]
Approving assuming comments are handled - consider me not blocking on this PR anymore

File: private/buf/bufcli/bufcli.go
Comment (joshcarp) [reply private/buf/bufcli/bufcli.go:829]:
```
// ImageFilteredByTypes returns a minimal image containing only the descriptors
// required to define those types. The resulting contains only files in which
// those descriptors and their transitive closure of required descriptors, with
// each file only contains the minimal required types and imports.
```
so yes

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (joshcarp) [reply private/buf/cmd/buf/command/beta/convert/convert.go:154]: "." will default to the current input directory. i.e. if you had a `buf.yaml` in your current directory you could run `buf convert --in=foo.json --out=foo.bin`

File: private/buf/cmd/buf/command/beta/convert/convert.go
Comment (joshcarp) [reply private/buf/cmd/buf/command/beta/convert/convert.go:171]: correct, that's handled at a much lower level by the compilers source accessor which has access to all of the wkts

**Pull request merged**
